diff --git a/ports/webassembly/main.c b/ports/webassembly/main.c
index fdd8ec5..637d0ea 100644
--- a/ports/webassembly/main.c
+++ b/ports/webassembly/main.c
@@ -47,7 +47,9 @@
 // This counter tracks the current depth of calls into C code that originated
 // externally, ie from JavaScript.  When the counter is 0 that corresponds to
 // the top-level call into C.
+#if MICROPY_GC_SPLIT_HEAP_AUTO
 static size_t external_call_depth = 0;
+#endif
 
 // Emscripten defaults to a 64k C-stack, so our limit should be less than that.
 #define CSTACK_SIZE (32 * 1024)
@@ -57,13 +59,15 @@ static void gc_collect_top_level(mp_obj_t root_obj);
 #endif
 
 void external_call_depth_inc(void) {
+    #if MICROPY_GC_SPLIT_HEAP_AUTO
     ++external_call_depth;
+    #endif
 }
 
 void external_call_depth_dec(mp_obj_t root_obj) {
+    #if MICROPY_GC_SPLIT_HEAP_AUTO
     --external_call_depth;
-    #if MICROPY_GC_SPLIT_HEAP_AUTO
     if (external_call_depth == 0) {
         gc_collect_top_level(root_obj);
     }
     #endif
 }
diff --git a/ports/webassembly/api.js b/ports/webassembly/api.js
index 4ef858f..383e294 100644
--- a/ports/webassembly/api.js
+++ b/ports/webassembly/api.js
@@ -152,15 +152,20 @@ const loadMicroPython = async ({
             const len = Module.lengthBytesUTF8(code);
             const buf = Module._malloc(len + 1);
             Module.stringToUTF8(code, buf, len + 1);
             const value = Module._malloc(3 * 4);
-            Module.ccall(
+            return Promise.resolve(Module.ccall(
                 "mp_js_do_exec_async",
                 "number",
                 ["pointer", "number", "pointer"],
                 [buf, len, value],
-            );
-            Module._free(buf);
-            const ret = proxy_convert_mp_to_js_obj_jsside_with_free(value);
-            if (ret instanceof PyProxyThenable) {
-                return Promise.resolve(ret);
-            }
-            return ret;
+                { async: true },
+            )).then(() => {
+                Module._free(buf);
+                const ret = proxy_convert_mp_to_js_obj_jsside_with_free(value);
+                if (ret instanceof PyProxyThenable) {
+                    return Promise.resolve(ret);
+                }
+                return ret;
+            }, (error) => {
+                Module._free(buf);
+                Module._free(value);
+                throw error;
+            });
         },
diff --git a/ports/webassembly/library.js b/ports/webassembly/library.js
index 1c390c5..4c0d172 100644
--- a/ports/webassembly/library.js
+++ b/ports/webassembly/library.js
@@ -34,8 +34,8 @@ mergeInto(LibraryManager.library, {
             const mp_interrupt_char = Module.ccall(
                 "mp_hal_get_interrupt_char",
                 "number",
-                ["number"],
-                ["null"],
+                [],
+                [],
             );
             const fs = require("fs");
